src/transformers/models/speecht5/modeling_speecht5.py

```diff
@@ -2587,36 +2595,71 @@ def _generate_speech(
         if output_cross_attentions:
             cross_attentions.append(torch.cat(decoder_out.cross_attentions, dim=0))
 
-        last_decoder_output = decoder_out.last_hidden_state[0, -1]
+        last_decoder_output = decoder_out.last_hidden_state.squeeze(1)
         past_key_values = decoder_out.past_key_values
 
         # Predict the new mel spectrum for this step in the sequence.
         spectrum = model.speech_decoder_postnet.feat_out(last_decoder_output)
-        spectrum = spectrum.view(model.config.reduction_factor, model.config.num_mel_bins)
+        spectrum = spectrum.view(bsz, model.config.reduction_factor, model.config.num_mel_bins)
         spectrogram.append(spectrum)
 
         # Extend the output sequence with the new mel spectrum.
-        output_sequence = torch.cat((output_sequence, spectrum[-1].view(1, 1, model.config.num_mel_bins)), dim=1)
-
+        output_sequence = torch.cat(
+            (output_sequence, spectrum[:, -1, :].view(bsz, 1, model.config.num_mel_bins)), dim=1
+        )
         # Predict the probability that this is the stop token.
         prob = torch.sigmoid(model.speech_decoder_postnet.prob_out(last_decoder_output))
 
-        # Finished when stop token or maximum length is reached.
-        if idx >= minlen and (int(sum(prob >= threshold)) > 0 or idx >= maxlen):
-            spectrogram = torch.cat(spectrogram, dim=0).unsqueeze(0)
-            spectrogram = model.speech_decoder_postnet.postnet(spectrogram)
-            spectrogram = spectrogram.squeeze(0)
-            break
-
-    if vocoder is not None:
-        outputs = vocoder(spectrogram)
+        if idx < minlen:
+            continue
+        else:
+            spectrograms = torch.stack(spectrogram)
+            spectrograms = spectrograms.transpose(0, 1).flatten(1, 2)
+            spectrograms = model.speech_decoder_postnet.postnet(spectrograms)
+
+            meet_thresholds = torch.sum(prob, dim=-1) >= threshold
+            # If the generation loops is less than maximum length times, check the ones in the batch that have met
+            # the prob threshold. Otherwise, assume all have met thresholds and fill other spectrograms for the batch.
+            meet_indexes = torch.where(meet_thresholds)[0] if idx < maxlen else torch.tensor(list(range(len(prob))))
+            meet_indexes = meet_indexes.tolist()
+            for meet_index in meet_indexes:
+                if meet_index in result_spectrogram:
+                    continue
+                result_spectrogram[meet_index] = spectrograms[meet_index]
+
```

Comment:
You could probably avoid entering this computationally heavy loop for a few iterations with fixes BTW.

Notice that you only need `prob`, `threshold`, `idx` and `result_spectrogram.keys()` to know if there is a need to change `result_spectrogram` at a given iteration. Knowing that, you can avoid computing `spectrograms` many times.

Let me know if you need help on this!

Reply:
The heavy loop means the `while True:` or the `for meet_index in meet_indexes:` one?

Reply:
Different spectrograms have different loops to meet the thresholds. What you mean is that some spectrograms that meet thresholds earlier do not need to be computed in further loops? Is what I mean correct?

Reply:
I actually was talking about entering the `else` condition. Notice that you recompute that snippet everytime, once `idx>=min_len`:

```python
            spectrograms = torch.stack(spectrogram)
            spectrograms = spectrograms.transpose(0, 1).flatten(1, 2)
            spectrograms = model.speech_decoder_postnet.postnet(spectrograms)
```

Whereas you only need to compute it when `meet_indexes` is non empty.

Reply:
Ahh yes, thanks for pointing that out:)